spec/compromise-syntax.md

```diff
@@ -0,0 +1,238 @@
+That also helps with embedding messages in various resource file formats,
+because they can freely trim surrounding white space without
+requiring escapes when a message pattern wants to start or end with spaces.
+
+By contrast, consider the experience with the existing ICU MessageFormat syntax
+which does start in “text mode”.
+ICU MessageFormat has pioneered the selection among multiple patterns based on run-time arguments.
+It represents selection using complex placeholders,
+which has the side effect of allowing literal text and other placeholders
+before and after the top-level selection placeholder.
+However, for reliable translations,
+there should be no translatable contents before or after the selection placeholder;
+instead, each selectable pattern should form one complete “translation unit”.
+Because the existing ICU MessageFormat starts in “text mode”,
+even though it looks like there is no extraneous text,
+spurious white space creeps in from developers’ line breaking of long message strings.
+The remedy is to always use syntax to indicate the start of translatable contents.
+
+We use curly braces to delimit patterns because
+`{}` are the paired ASCII punctuation characters least commonly used in normal text.
+For the same reason, we also use them for embedding placeholders in patterns.
+
+Literal text can use any characters except for curly braces,
+and except for the backslash, which we use as usual for escaping.
+That is, the only special characters inside a pattern are `{}\`.
+The only allowed escape sequences are `\{`, `\}`, and `\\`.
+It is an error if `\` is followed by any other character.
+
+The message syntax does not use `'` or `"`,
+so that it is easy to hard-code message strings in programming language source code.
+
+# Placeholders
+
+Formatting a message replaces placeholders with values based on run-time arguments or special functions.
+We also allow for value literals specified inside the placeholder,
+instead of using an argument name;
+and we also allow for invoking functions without using argument names or value literals.
+```
+{$name}
+{$count :number}
+{$fraction :number style=percent minFractions=2}
+{<25> :number}
+{:specialFunction optionKey=optionValue key2=<value with spaces>}
+```
+
+An argument name is a `$` immediately followed by an identifier.
+A message formatting function will typically accept a Map of argument keys to values
+where the keys match argument name identifiers in the patterns of the message.
+
+TODO: For the definition of identifiers we should consult with the Unicode Source Code Working Group.
+
+If the placeholder specifies only an argument name,
+then the formatting function is inferred from the run-time type of the argument value.
+For example, a string value would simply be inserted,
+and a numeric type could be formatted using some kind of default number formatter.
+- TODO: In the registry, specify the default formatters for a small set of value types.
```

Comment:
Perhaps this is upside down. The registry should specify a set of formatters (to which an implementation can add) and these can "register" what types they service (and in what priority order). At Amazon our message formatter has a currency formatter function (`PriceFormat`) that handles `Price` objects--the `Price` object extends `Number`, but `PriceFormat` takes priority for that type.

Reply:
With “the registry” I mean the future CLDR file that defines functions with their names, options, and semantics. That should include what formatter to use for a numeric argument when no function is explicitly specified in the message. This registry could specify a different formatter for a subtype.

It sounds like what you are referring to would be some runtime object that can dynamically handle types and formatters. I think that's out of scope for this document.